decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java

```diff
@@ -0,0 +1,305 @@
+    private final List<Slice<T, L>> slices;
+    private final Map<String, GraphEdge> edges;
+    private int currentSourceSliceIndex = -1;
+    private GraphEdge<T, L> graphEdge = null;
+    private GraphEdge<T, L> lastSuccessfulEdge = null;
+
+    public BackTrackingGraph(List<Slice<T, L>> slices) {
+        this.slices = slices;
+        this.edges = new HashMap<>();
+    }
+
+    /**
+     * if the user has marked last edge returned by the graph as red;
+     * then the next edge should be an unmarked edge with the same start node
+     * But If all the edges starts from this start node is already marked
+     * then the next edge can start from any node in the slice
+     *
+     * @param sourceSlice Slice from which next edge will start if all edges from the start node of previous edge is already marked
+     * @return index of the node from which next edge starts
+     */
+    private int findNextNodeIndexIfCurrentIsRed(Slice sourceSlice) {
+        GraphNode<T, L> sourceNode = graphEdge.getSourceNode();
+        if (sourceNode.isDeadEnd() || !sourceNode.hasUnmarkedEdges()) {
+            return sourceSlice.getNodePosition(sourceSlice.getNextBestNode());
+        } else {
+            return sourceSlice.getNodePosition(graphEdge.getSourceNode());
+        }
+    }
+
+    /**
+     * Index of the start node of next best edge from the slice
+     *
+     * @param sourceSlice
+     * @return index of the node from which next edge starts
+     */
+    private int getNextSourceNodeIndex(Slice sourceSlice) {
+        if (graphEdge == null) {
+            GraphNode sourceNode = sourceSlice.getNextBestNode();
+            return sourceSlice.getNodePosition(sourceNode);
+        } else if (graphEdge.getEdgeColor().equals(EntityColor.RED)) {
+            handleIncomingEdges(graphEdge.getSourceNode(), sourceSlice);
+            return findNextNodeIndexIfCurrentIsRed(sourceSlice);
+        } else {
+            return sourceSlice.getNodePosition(graphEdge.getDestinationNode());
+        }
+
+    }
+
+    /**
+     * We don't have to explore any incoming edges if node is a dead end
+     *
+     * @param sourceNode
+     * @param sourceSlice
+     */
+    private void handleIncomingEdges(GraphNode<T, L> sourceNode, Slice sourceSlice) {
+        if (sourceNode.isDeadEnd() && slices.indexOf(sourceSlice) != slices.size() - 1) {
+            sourceNode.getIncomingEdge().forEach(
+                    edge -> {
+                        edge.setEdgeColor(EntityColor.RED);
+                    }
+            );
+        }
+    }
+
+    /**
+     * Backtracking is required if the green edges in the previous slices are not connected directly or indirectly to the given edge
+     *
+     * @param currentSliceIndex index of the slice from which the edge starts
+     * @param graphEdge         edge of which the predecessors need to verified
+     * @return
+     */
+    public boolean isBacktrackingRequired(int currentSliceIndex, GraphEdge<T, L> graphEdge) {
```

Comment:
Does this method need to be `public`?  Is the client going to need to call it?  I don't see any uses of it outside of the class itself.